### PR TITLE
Use FA's CrossEntropyLoss for metrics calculation

### PR DESCRIPTION
**Pull request by snarayan21**
# What does this PR do?

If available, we should use FA's CrossEntropyLoss for calculating the CE metric in composer. We already use it in foundry, and it's simply faster. Below we can see that the CE loss metric is the same, and there's a nice MFU boost as well.

Pink: using FA's CrossEntropyLoss
Green: using torch's CrossEntropyLoss
Screenshot 2024-04-25 at 4 47 49 PM
Screenshot 2024-04-25 at 4 48 41 PM

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by dakinggg** [comment]
https://github.com/mosaicml/composer/pull/2987 this is the third time we've had this PR :joy: once in foundry, once in composer, and now again. @mvpatel2000 last time I thought we should put it in foundry. Do you want to take FA as a Composer dependency?

**Review by mvpatel2000 — COMMENTED** [review]
Yea Im fine adding it. @snarayan21  please add as an optional dependency

**Comment by snarayan21** [comment]
added as optional dep

**Review by b-chu — APPROVED** [review]
Thanks for the PR!

File: setup.py [hunk 1]
```diff
@@ -225,6 +225,10 @@ def package_files(prefix: str, directory: str, extension: str):
     'mlflow>=2.11.1,<3.0',
 ]
 
+extra_deps['gpu-flash2'] = [
+    'flash-attn==2.5.0',
+]
```
Comment (mvpatel2000) [inline setup.py:230]: why `gpu-` instead of jsut `flash2`?

Comment (dakinggg) [reply setup.py:230]: should just be `flash-attention`

Comment (snarayan21) [reply setup.py]: wbout `flash-attn` since its the package name

Comment (snarayan21) [reply setup.py]: changed to `flash-attn`

Comment (dakinggg) [reply setup.py]: ya sounds good

File: pyproject.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # build requirements
 [build-system]
-requires = ["setuptools < 68.0.0"]
+requires = ["setuptools < 68.0.0", "packaging >= 21.3.0, < 24.1"]
```
Comment (mvpatel2000) [inline pyproject.toml:3]: why this?

Comment (snarayan21) [reply pyproject.toml:3]: gets around some of the installation errors, but torch is also a dependency, and then we go round and round in this circle trying to get build isolation working correctly. so i'm closing this PR, it should live foundry side.

**Comment by snarayan21** [comment]
Closed this PR since adding `flash-attn` to composer as a dependency makes things very ugly. Will add this to foundry later. 4th time's the charm!

**Pull request closed without merging**